W 10673773 apicatalog updates for ga bsj (#189)
mulesoft/docs-exchange: modules/ROOT/pages/apicat-publish-using-api-catalog-cli.adoc

Create a descriptor file that describes the API definitions to publish and copy the file to the file path where your API projects are stored. See xref:apicat-create-descriptor-file-cli#create-desc-file-cli[Create a Descriptor File Using the CLI].

To publish an API definition using the CLI:

* Run the `api-catalog publish-asset` command. The following example publishes API definitions described in the `catalog.yaml` file:
+
----
> api-catalog publish-asset -a https://anypoint.mulesoft.com -d catalog.yaml -o 1234abc-b34a-7cd3-4s32-12345abc2345 -u myAnyPtAccount -p myPwd@4!myacct 
----

//include::anypoint-cli::partial$api-catalog.adoc[tag=publish-asset-command]

include::partial$api-catalog.adoc[tag=publish-asset-command]

The following sections give example commands that show how several catalog options are used:

* <<default-settings,Publish an Asset with Default Settings>>

* <<dry-run,Perform a Dry Run to Validate the Descriptor File>>

* <<force-publish,Publish an Asset Regardless of the Content>>

* <<trigger-conditions,Trigger the Publish Using Conditions>>

* <<jenkins-script,Publish Using a Jenkins Script>> 

[[default-settings]]
== Publish an Asset with Default Settings

> The following example publishes the API to an organization in the US Anypoint Platform control plane. Defaults or environment variable settings are used for the unspecified options. 

Reviewer: ```suggestion
The following example publishes the API to an organization in the US control plane for Anypoint Platform. Defaults or environment variable settings are used for the unspecified options. 
```

Or 

```suggestion
The following example publishes the API to an organization in the Anypoint Platform US control plane. Defaults or environment variable settings are used for the unspecified options. 
```

----
> api-catalog publish-asset -o 1234abc-b34a-7cd3-4s32-12345abc2345  -a https://anypoint.mulesoft.com/

----

This example assumes that either the username and password credentials are defined in the ANYPOINT_USERNAME and ANYPOINT_PASSWORD environment variables or that the client ID and client secret credentials are already defined in ANYPOINT_CLIENT_ID and ANYPOINT_CLIENT_SECRET environment variables. 

[[dry-run]]
== Perform a Dry Run to Validate the Descriptor File

This example shows the `api-catalog publish-asset` command with the following options:

> * --dry-run: Runs the command to verify that the descriptor file is valid but does not publish the APIs

Reviewer: ```suggestion
* --dry-run: Runs the command to verify that the descriptor file is valid but does not publish the APIs.
```

> * --verbose: Shows the maximum information in the results

Reviewer: ```suggestion
* --verbose: Shows the maximum information in the results.
```

The example command sends the standard output to `mycatdryrun.log`. 

----
> api-catalog publish-asset -a https://anypoint.mulesoft.com -d catalog.yaml -o 1234abc-b34a-7cd3-4s32-12345abc2345 --client-id=1234567abcd2345gabc987656abc --client-secret --dry-run --verbose >> mycatdryrun.log

----

[[force-publish]]
== Publish an Asset Regardless of the Content

To publish a new version of the asset in Exchange regardless of the content, use the `--force-publish` option. 

This example shows the `api-catalog publish-asset` command with both of these options. This example assumes that authentication credentials are set in environment variables.

----
> api-catalog publish-asset -a https://anypoint.mulesoft.com --descriptor-file=mydescriptor.yaml -o 1234abc-b34a-7cd3-4s32-12345abc2345 --force-publish 

----

[[trigger-conditions]]

> == Trigger the Publish Using Conditions 

Reviewer: "Trigger the Publish" seems incorrect. Is there some other way to phrase this?

> Custom trigger conditions enable you to control whether APIs are published when the API Catalog CLI command runs. This example shows how `--trigger-criteria` options correlate with descriptor file trigger conditions to control whether or not to publish the described APIs.   

Reviewer: ```suggestion
Custom trigger conditions enable you to control whether APIs are published when you run API Catalog CLI commands. This example shows how `--trigger-criteria` options correlate with descriptor file trigger conditions to control whether or not to publish the described APIs.   
```

> *Example Trigger Conditions in a Descriptor file*

Reviewer: ```suggestion
*Example trigger conditions in a descriptor file*
```

> Following is an example `triggerConditions` section in the `mydescriptor.yaml` descriptor file at the global level:

Reviewer: ```suggestion
Example `triggerConditions` section in the `mydescriptor.yaml` descriptor file at the global level:
```

---
 
 triggerConditions:
  ref:
   - main
  tags:
   - migrate
   - republish
  status:
   - complete
   - ready

---

> *Example That Triggers the Publish*

Reviewer: ```suggestion
*Example that triggers the publish*
```

Reviewer: again, "the publish"? just seems strange.

> The following command triggers the API publish because the trigger option values match all of the global trigger conditions in the `mydescriptor.yaml` descriptor file. 

Reviewer: ditto

----
> api-catalog publish-asset -a https://anypoint.mulesoft.com -d mydescriptor.yaml -o 1234abc-b34a-7cd3-4s32-12345abc2345 -u myAnyPtAccount -p myPwd@4!myacct --trigger-criteria=ref:main --trigger-criteria=tags:migrate --trigger-criteria=status:complete

----

> *Example That Does Not Trigger the Publish*

Reviewer: ```suggestion
*Example that does not trigger the publish*
```

Reviewer: ditto

> The following command does not trigger the API publish because the trigger option value for `ref` is `test`, not `main`. Because not all trigger values match the corresponding global trigger conditions, no APIs are published. 

Reviewer: ditto

----
> api-catalog publish-asset -a https://anypoint.mulesoft.com -d mydescriptor.yaml -o 1234abc-b34a-7cd3-4s32-12345abc2345 -u myAnyPtAccount -p myPwd@4!myacct --trigger-criteria=ref:test --trigger-criteria=tags:migrate --trigger-criteria=status:complete

----

[[versionStrategyConditions]]
== Set Version Strategy Using Conditions 

> Version strategy conditions enable you to set the version of assets when the API Catalog CLI publish command runs as part of your CI/CD pipeline or scripts. This example shows how `--version-strategy-criteria` options correlate with descriptor file version strategy conditions to set the version strategy for the described APIs. If the version strategy is set at the global level, it overrides values set at any other level. If the value for a condition matches more than one version strategy, the command gives an error.

Reviewer: oh wait. It's the publish command??? should it be the `publish` command?

> Version strategy conditions enable you to set the version of assets when the API Catalog CLI publish command runs as part of your CI/CD pipeline or scripts. This example shows how `--version-strategy-criteria` options correlate with descriptor file version strategy conditions to set the version strategy for the described APIs. If the version strategy is set at the global level, it overrides values set at any other level. If the value for a condition matches more than one version strategy, the command gives an error.

Reviewer: ```suggestion
Version strategy conditions enable you to set the version of assets when you run the API Catalog CLI `publish` command as part of your CI/CD pipeline or scripts. This example shows how `--version-strategy-criteria` options correlate with descriptor file version strategy conditions to set the version strategy for the described APIs. If you set the version strategy at the global level, it overrides values set at any other level. If the value for a condition matches more than one version strategy, the command gives an error.
```

> *Example Version Strategy Conditions in a Descriptor file*

Reviewer: ```suggestion
*Example version strategy conditions in a descriptor file*
```

> Following is an example `versionStrategyConditions` section in the `mydescriptor.yaml` descriptor file. 

Reviewer: ```suggestion
Example `versionStrategyConditions` section in the `mydescriptor.yaml` descriptor file. 
```

----
 
versionStrategyConditions:
patchIncrease:
branch:
 ‘main’
author:
 ‘admin’
snapshot:
branch:
‘test’
‘Develop’

----

> *Example That Sets the Version Strategy During the Publish*

Reviewer: ```suggestion
*Example that sets the version strategy during the publish*
```

Reviewer: So could this actually be "...the version strategy when you run the `publish` command."?

> The following command sets multiple version strategies during the publish because the option values match multiple conditions in the `mydescriptor.yaml` descriptor file. 

Reviewer: ditto

----
> api-catalog publish-asset -a https://anypoint.mulesoft.com -d mydescriptor.yaml -o 1234abc-b34a-7cd3-4s32-12345abc2345 -u myAnyPtAccount -p myPwd@4!myacct ----version-strategy-criteria=branch:main ----version-strategy-criteria=author:admin ----version-strategy-criteria=branch:test ----version-strategy-criteria=branch:Develop
----

The resulting version strategies are:

* Patch increase for APIs in the `main` branch with an author of `admin`. 
* Snapshot for APIs in the `test` and `Develop`branches. 

> *Example That Does Not Set the Version Strategy During the Publish*

Reviewer: ```suggestion
*Example that does not set the version strategy during the publish*
```

Reviewer: Sorry, but now it's just dawned on me, should these be headings?
